File: parser/src/main/java/dev/mvmo/sicklang/parser/ast/expression/IndexExpressionNode.java
package dev.mvmo.sicklang.parser.ast.expression;

import dev.mvmo.sicklang.token.Token;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor(staticName = "newInstance")
public class IndexExpressionNode implements ExpressionNode {

    private final Token token;
    private final ExpressionNode left;
    private final ExpressionNode right;

    @Override
    public String tokenLiteral() {
        return token.literal();
    }

    @Override
    public String toString() {
        return String.format("(%s[%s])", left.toString(), right.toString());
    }

}

File: parser/src/main/java/dev/mvmo/sicklang/parser/ast/expression/IndexExpressionNode.kt
package dev.mvmo.sicklang.parser.ast.expression

import dev.mvmo.sicklang.token.Token

class IndexExpressionNode(val token: Token, val left: ExpressionNode, val index: ExpressionNode) : ExpressionNode {

    override fun tokenLiteral(): String =
        token.literal

    override fun toString(): String =
        "($left[$index])"

}